src/otx/algo/samplers/balanced_sampler.py

```diff
@@ -0,0 +1,160 @@
+            tail of the data to make it evenly divisible across the number of
+            replicas. If ``False``, the sampler will add extra indices to make
+            the data evenly divisible across the replicas. Default: ``False``.
+        n_repeats (Union[float, int, str], optional) : number of iterations for manual setting
+    """
+
+    def __init__(
+        self,
+        dataset: OTXDataset,
+        samples_per_gpu: int,
+        efficient_mode: bool = False,
+        num_replicas: int = 1,
+        rank: int = 0,
+        drop_last: bool = False,
+        n_repeats: float | int | str = 1,
+    ):
+        self.samples_per_gpu = samples_per_gpu
+        self.num_replicas = num_replicas
+        self.rank = rank
+        self.drop_last = drop_last
+
+        super().__init__(dataset, samples_per_gpu, n_repeats=n_repeats)
+
+        # img_indices: dict[label: list[idx]]
+        ann_stats = compute_class_statistics(dataset.dm_subset)
+        self.img_indices = {k: v for k, v in ann_stats.items() if len(v) > 0}
+        self.num_cls = len(self.img_indices.keys())
+        self.data_length = len(self.dataset)
+        self.num_trials = int(self.data_length / self.num_cls)
+
+        if efficient_mode:
+            # Reduce the # of sampling (sampling data for a single epoch)
+            num_tail = min(len(cls_indices) for cls_indices in self.img_indices.values())
+            if num_tail > 1:
+                base = 1 - (1 / num_tail)
+                num_reduced_trials = int(math.log(0.001, base))
+                self.num_trials = min(num_reduced_trials, self.num_trials)
+
+        self.num_samples = self._calculate_num_samples()
+
+    def _calculate_num_samples(self) -> int:
+        num_samples = self.num_trials * self.num_cls * self.repeat
+
+        if self.num_replicas > 1:
+            # If the dataset length is evenly divisible by # of replicas, then there
+            # is no need to drop any data, since the dataset will be split equally.
+            if self.drop_last and num_samples % self.num_replicas != 0:
+                # Split to nearest available length that is evenly divisible.
+                # This is to ensure each rank receives the same amount of data when
+                # using this Sampler.
+                num_samples = math.ceil(
+                    # `type:ignore` is required because Dataset cannot provide a default __len__
+                    # see NOTE in pytorch/torch/utils/data/sampler.py
+                    (num_samples - self.num_replicas) / self.num_replicas,
+                )
+            else:
+                num_samples = math.ceil(num_samples / self.num_replicas)
+            self.total_size = num_samples * self.num_replicas
+
+        return num_samples
+
+    def __iter__(self):
+        """Iter."""
+        indices = []
+        random_generator = np.random.default_rng()
+        for _ in range(self.repeat):
```

Comment:
See this link https://pytorch.org/docs/2.1/_modules/torch/utils/data/sampler.html#RandomSampler and I hope you follow a flow as similar to the PyTorch codebase as possible, especially for the random generator part.

Reply:
I replaced `np.Generator` with `torch.Generator` and modified the syntax around generator accordingly. - https://github.com/openvinotoolkit/training_extensions/pull/2993/commits/f09c0192faf1045647991bfdf24c426b5eeb5585